### PR TITLE
fix: support project finalizer to ensure proper deletion

### LINKED ISSUE #1329: Deadlock on deleting application
During experiments i found the following "deadlock" when deleting an application. 

I have an application which provisions an appproject a namespace and a new application (which is in the new project and deloys services in the new namespace) 

image

Doing a cascading delete on test-app yields the following result: 

the local-joe application stays behind with the error message: 
InvalidSpecError
Application referencing project local-joe which does not exist

Application test-app stays behind with the error: 
DeletionError
Operation cannot be fulfilled on namespaces "local-joe": The system is ensuring all content is removed from this namespace. Upon completion, this namespace will automatically be purged by the system.

What inspecting the cluster shows is that all the resources from app local joe are gone. 
The namespace is deleted

A way to fix this behavior is to manually issue: 
```
kubectl delete apps local-joe
application.argoproj.io "local-joe" deleted
```
which removes the local-joe app and allows test-app to delete. 

Also attached the application controller log. 
[delete.log](https://github.com/argoproj/argo-cd/files/3014333/delete.log)

### LINKED ISSUE #3175: bug: deleting project and application at the same time using autosync&prune
Checklist:

* [x] I've searched in the docs and FAQ for my answer: http://bit.ly/argocd-faq.
* [x] I've included steps to reproduce the bug.
* [x] I've pasted the output of `argocd version`.

**Describe the bug**

In an environment with the following conditions: 
- ArgoCD is using declarative setup in an apps of apps style
- all apps are set to auto-sync including auto-prune

I wanted to delete one of the projects and the applications using the project. So I deleted the files defining the approject and the application in a single commit. 

ArgoCD picked up the diff and started the sync and successfully deleted the project. For the deletion of the application, ArgoCD wants to prune all objects now, but the project is deleted already, Therefore ArgoCD stopped to delete the objects, because there is no project matching the application.

**To Reproduce**

- create a project and an application using this project with autosync and prune enabled
- create an app maintaining this project and application using autosync and autoprune
- commit and push to repo
- wait for the sync to create all objects
- delete the new project and application from repo
- commit and push
-> project gets deleted, application is in unknown state trying to delete its objects

**Expected behavior**

ArgoCD waits for the deletion of the application (including its maintained objects) before the project gets deleted.

**Screenshots**

**Version**

```shell
argocd-server: v1.4.2+unknown
  BuildDate: 2020-02-03T11:45:44Z
  GitCommit:
  GitTreeState: clean
  GoVersion: go1.13.5
  Compiler: gc
  Platform: linux/amd64
  Ksonnet Version: v0.13.1
  Kustomize Version: Version: {Version:kustomize/v3.2.3 GitCommit:f8412aa3d39f32151525aff97a351288f5a7470b BuildDate:2019-10-08T23:30:25Z GoOs:linux GoArch:amd64}
  Helm Version: v2.15.2
  Kubectl Version: v1.14.0
```

**Logs**

```
ComparisonError     appproject.argoproj.io "deletebug-test" not found
DeletionError       appproject.argoproj.io "deletebug-test" not found
InvalidSpecError    Application referencing project deletebug-test which does not exist
```

### PR DESCRIPTION
**Pull request by alexmt**
Closes https://github.com/argoproj/argo-cd/issues/1329
Closes https://github.com/argoproj/argo-cd/issues/3175

PR adds deletion finalizer to the appproj resource which ensures that declaratively managed projects are deleted only after all applications are deleted.

### CODE REVIEW

File: controller/appcontroller.go [hunk 1]
```diff
@@ -559,6 +585,71 @@ func (ctrl *ApplicationController) processAppComparisonTypeQueueItem() (processN
...
+		var patch []byte
+		patch, _ = json.Marshal(map[string]interface{}{
+			"metadata": map[string]interface{}{
+				"finalizers": proj.Finalizers,
+			},
+		})
+		_, err = ctrl.applicationClientset.ArgoprojV1alpha1().AppProjects(ctrl.namespace).Patch(proj.Name, types.MergePatchType, patch)
+		if err != nil {
+			return err
+		}
```
Comment (darshanime) [inline controller/appcontroller.go:648]: can this be made into a function - maybe in a separate PR? 
Almost always, whereever we use `RemoveFinalizer`, we follow it up with `Patch` on api-server.

File: controller/appcontroller.go [hunk 2]
```diff
@@ -559,6 +585,71 @@ func (ctrl *ApplicationController) processAppComparisonTypeQueueItem() (processN
...
+	}
+	return nil
```
Comment (darshanime) [inline controller/appcontroller.go:650]: maybe a log line to say `Cannot remove finalizer as project is referenced by ... applications` if `hasApps` is true?

File: docs/operator-manual/project.yaml [hunk 3]
```diff
@@ -3,6 +3,9 @@ kind: AppProject
 metadata:
   name: my-project
   namespace: argocd
+  # Finalizer that ensures that project is not deleted until it is referenced by any application
```
Comment (mayzhang2000) [inline docs/operator-manual/project.yaml:6]: Should it be "Finalizer that ensures that project is not deleted until it is _Not_ referenced by any application"

File: controller/appcontroller.go (hunk 1 above)
Comment (alexmt) [reply controller/appcontroller.go]: makes sense. Moved code into separate method

File: controller/appcontroller.go (hunk 2 above)
Comment (alexmt) [reply controller/appcontroller.go:642]: 👍 , added

File: docs/operator-manual/project.yaml (hunk 3 above)
Comment (alexmt) [reply docs/operator-manual/project.yaml]: nice catch. fixed

File: docs/operator-manual/declarative-setup.md [hunk 4]
```diff
@@ -84,6 +84,9 @@ kind: AppProject
 metadata:
   name: my-project
   namespace: argocd
+  # Finalizer that ensures that project is not deleted until it is referenced by any application
```
Comment (mayzhang2000) [inline docs/operator-manual/declarative-setup.md:87]: Same here: until it is not referenced by any application

**Comment by alexmt** [comment]
Thank you for review @darshanime and @mayzhang2000 !

**Review by mayzhang2000 — APPROVED** [review]

File: docs/operator-manual/declarative-setup.md (hunk 4 above)
Comment (alexmt) [reply docs/operator-manual/declarative-setup.md]: done

**Pull request merged**